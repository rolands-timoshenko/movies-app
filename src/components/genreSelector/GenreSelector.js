import React from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import { Col, Button } from 'reactstrap';

const GenreSelector = observer(({genres, onChange}) => {
  
  const buttonStyle = {
    marginBottom:20, 
    width: '100%'
  };

  return genres.map((genre) => {
    return (
      <Col lg="2" md="3" sm="4" xs="6" key={genre.id}>
        <Button color={genre.isActive ? "primary": "secondary"} active={genre.isActive} style={buttonStyle} onClick={onChange(genre.id)}>{genre.name}</Button>
      </Col>
    );
  });
});

GenreSelector.propTypes = {
  genres: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired
};

export { GenreSelector };