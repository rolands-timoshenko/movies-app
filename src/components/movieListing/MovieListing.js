import React from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import { MovieCard } from './movieCard/MovieCard';
import { Col } from 'reactstrap';

const MovieListing = observer(({movies}) => {
  return movies.map( movie => (
      <Col key={movie.id} xs="12" sm="6" md="4" lg="3">
        <MovieCard movie={movie} />
      </Col>
  ) );
});

MovieListing.propTypes = {
  movies: PropTypes.array.isRequired
};

export { MovieListing }