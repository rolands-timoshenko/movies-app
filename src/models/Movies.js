import { types } from 'mobx-state-tree';
import MoviesApi from '../api/MoviesApi';

const Genres = types.model("Genres", {
  id: types.number,
  name: types.string
});

// Movie model
const Movie = types.model("Movie", {
  id: types.number,
  title: types.string,
  posterPath: types.maybeNull(types.string),
  popularity: types.number,
  voteAverage: types.number
})
.views(self => ({
  get posterImage() {
    return `${MoviesApi.imageUrl}${self.posterPath}`;
  }
}));

// Main store model
const MovieStore = types.model("Movies", {
  totalPages: types.maybe(types.number),
  currentPage: types.maybe(types.number),
  movies: types.array(Movie),
  genres: types.array(Genres)
})

// Computed values
.views(self => ({
  get moviesByPopularity() {
    return self.movies.sort((a, b) => {
      return b.popularity - a.popularity;
    });
  }
}))
// Actions are for interacting with store
.actions(self => {

  // Set api instance
  const api = new MoviesApi();

  // Return public methods
  return {

    async loadMovies(page = 1) {
      self.movies = [];
      const data = await api.getPlayingMovies(page);
      self.setTotalPages(data.total_pages);
      self.setMovies(data.results);
      self.setCurrentPage(data.page);
    },

    async loadGenres() {
      const genres = await api.getMovieGenres();
      self.setGenres(genres);
    },

    loadNext() {
      self.loadMovies(self.currentPage + 1);
    },

    loadPrev() {
      self.loadMovies(self.currentPage - 1);
    },

    setTotalPages(pages) {
      self.totalPages = pages;
    },

    setCurrentPage(page) {
      self.currentPage = page;
    },

    setMovies(movies) {
      movies.map((movie) => {
        self.movies.push(Movie.create({
          id: movie.id,
          title: movie.title,
          posterPath: movie.poster_path,
          popularity: movie.popularity,
          voteAverage: movie.vote_average
        }));
      });
    },

    setGenres(genres) {
      genres.map((genre) => {
        self.genres.push(Genres.create({
          id: genre.id,
          name: genre.name
        }));
      });
    }
  }
});

export default MovieStore;