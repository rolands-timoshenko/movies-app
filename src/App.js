import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import { observer, inject } from 'mobx-react';
import { RatingSlider } from './components/ratingSlider/RatingSlider';
import { GenreSelector } from './components/genreSelector/GenreSelector';
import { MovieListing } from './components/movieListing/MovieListing';
import { Button, Container, Row, Col } from 'reactstrap';

export class App extends Component {

  rootStore;

  constructor(props) {
    super(props);
    this.rootStore = this.props.rootStore;
  }

  componentDidMount() {
    // Lets load movies and genres 
    this.rootStore.loadMovies();
    this.rootStore.loadGenres();
  }

  // Show Load More button if there are more pages
  showLoadMore = () => {
    return (this.rootStore.totalPages > this.rootStore.currentPage)
      ? <Button color="primary" onClick={this.rootStore.loadNext}>Load More</Button> : null;
  }

  // On cmp render function will be called and id will be saved in closure
  handleSelectGenre = (id) => {
    // Returning onChange handler. Using preserve id
    return () => {
      this.rootStore.selectGenre(id);
    }
  }

  // On slider change lets update store
  handleOnSliderChange = (val) => {
    this.rootStore.setRating(val);
  }

  render() {

    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Welcome to Movie listing app</h1>
        </header>
        <Container>
          <Row>
            <RatingSlider onChange={this.handleOnSliderChange} value={this.rootStore.selectedRating} />
          </Row>
          <Row>
            <GenreSelector onChange={this.handleSelectGenre} genres={this.rootStore.genres} />
          </Row>
          <Row>
            <MovieListing movies={this.rootStore.moviesFilteredByrating} />
          </Row>
          <Row>
            <Col>
              {this.showLoadMore()}
            </Col>
          </Row>
        </Container>
      </div>
    );
  }
}

export default inject('rootStore')(observer(App));
