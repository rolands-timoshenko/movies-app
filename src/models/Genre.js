import { types, getParentOfType } from 'mobx-state-tree';
import { RootStore } from './RootStore';

export const Genre = types.model("Genre", {
  id: types.number,
  name: types.string
})

.views(self => ({
  get isActive() {
    const genres = getParentOfType(self, RootStore).selectedGenres;
    return genres.includes(self.id);
  }
}));