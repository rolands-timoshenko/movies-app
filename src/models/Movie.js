import { types, getParentOfType } from 'mobx-state-tree';
import MoviesApi from '../api/MoviesApi';
import { RootStore } from './RootStore';

export const Movie = types.model("Movie", {
  id: types.number,
  title: types.string,
  posterPath: types.maybeNull(types.string),
  popularity: types.number,
  voteAverage: types.number,
  genreIds: types.array(types.number)
})
// Computed values
.views(self => ({

  // Get genres as Genre[]
  get genres() {
    const genres = getParentOfType(self, RootStore).genres;
    return genres.filter((genre) => {
      return self.genreIds.includes(genre.id);
    });
  },

  // This one will give us genre names as string[]
  get genreNames() {
    return self.genres.map((genre) => {
      return genre.name
    })
  },

  // Lets get full image path. genre model will have posterImage property. 
  get posterImage() {
    return `${MoviesApi.imageUrl}${self.posterPath}`;
  }
}));