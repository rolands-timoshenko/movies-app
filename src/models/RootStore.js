import { types } from 'mobx-state-tree';
import MoviesApi from '../api/MoviesApi';
import { Movie } from './Movie';
import { Genre } from './Genre';

const ERROR = {
  NOTHING_TO_LOAD: 0
}

// Main store model
export const RootStore = types.model("Movies", {
  totalPages: types.maybe(types.number),
  currentPage: types.maybe(types.number),
  movies: types.array(Movie),
  genres: types.array(Genre),
  selectedGenres: types.array(types.number),
  selectedRating: types.optional(types.number, 3),
  loadingMovies: types.optional(types.boolean, false),
  nothingToLoad: types.optional(types.boolean, false)
})

// Computed values
.views(self => ({

  // Movies will be sorted by popularity. Lets use slice, so we are not modifying original one.
  get moviesByPopularity() {
    return self.movies.slice().sort((a, b) => {
      return b.popularity - a.popularity;
    });
  },

  // Movies will be filtered by genre if such selected
  get moviesByGenre() {
    if (self.selectedGenres.length > 0) {
      return self.moviesByPopularity.filter((movie) => {
        return self.selectedGenres.every((id) => {
          return movie.genreIds.includes(id);
        });
      }) 
    }
    return self.moviesByPopularity;
  },

  // Movies will be sorted by selected rating and higher
  get moviesFilteredByrating() {
    return self.moviesByGenre.sort((a, b) => b.voteAverage - a.voteAverage).filter((movie) => {
      return movie.voteAverage >= self.selectedRating;
    });
  },

  // Get actual genres. If only there is movie with such genre
  get actualGenres() {

    // Lets get all actual genre ids.
    // Later it will be used for filtering genres
    const genreIds = self.moviesFilteredByrating
      .map(movie => movie.genreIds)
      .reduce((col, curr) => [...new Set([...col, ...curr])], []);
    return self.genres.slice().filter((genre) => {
      return genreIds.includes(genre.id);
    });
  }

}))
// Actions are for interacting with store
.actions(self => {

  // Set api instance
  const api = new MoviesApi();

  function spreadData(data) {
    self.setTotalPages(data.total_pages);
    self.setMovies(data.results);
    self.setCurrentPage(data.page);
  }

  // Return public methods
  return {

    // This one will load movies from api and will spread properties to store
    async loadMovies(page = 1) {
      if (self.nothingToLoad) {
        return;
      }
      try {
        self.setLoadingMovies(true);
        const data = await api.getPlayingMovies(page);
        if (data.results.length < 1 || data.total_results === self.movies.length) {
          throw new Error(ERROR.NOTHING_TO_LOAD);
        }
        spreadData(data);
        self.setLoadingMovies(false);
      } catch (e) {
        
        switch (e.message) {
          case ERROR.NOTHING_TO_LOAD:
            self.setNothingToLoad(true);
            break;

          default:
            throw e;
        }
      } finally {
        self.setLoadingMovies(false);
      }
    },

    // Will load genres from api and spread them into store
    async loadGenres() {
      const genres = await api.getMovieGenres();
      self.setGenres(genres);
    },

    loadNext() {
      self.loadMovies(self.currentPage + 1);
    },

    setTotalPages(pages) {
      self.totalPages = pages;
    },

    setCurrentPage(page) {
      self.currentPage = page;
    },

    // movies will be pushed into store
    setMovies(movies) {
      movies.forEach((movie) => {
        // push new movie into store
        self.movies.push(Movie.create({
          id: movie.id,
          title: movie.title,
          posterPath: movie.poster_path,
          popularity: movie.popularity,
          voteAverage: movie.vote_average,
          genreIds: movie.genre_ids
        }));
      });
    },

    // Will set rating
    setRating(val) {
      if (self.selectedRating !== val) {
        self.selectedRating = val;
      }
    },

    // Saving genres inside store
    setGenres(genres) {
      genres.map((genre) => {
        return self.genres.push(Genre.create({
          id: genre.id,
          name: genre.name
        }));
      });
    },

    setNothingToLoad(bool) {
      self.nothingToLoad = bool;
    },

    setLoadingMovies(bool) {
      self.loadingMovies = bool;
    },

    // Will check if genre already exists in selectedGenres[] if so, item will be removed
    selectGenre(id) {
      if (self.selectedGenres.includes(id)) {
        const newSelectedGenres = self.selectedGenres.filter((genreId) => {
          return genreId !== id;
        });
        self.selectedGenres = [...newSelectedGenres];
      } else {
        self.selectedGenres.push(id);
      }
    }

  }
});