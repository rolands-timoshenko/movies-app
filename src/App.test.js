import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import { GenreSelector } from './components/genreSelector/GenreSelector';
import { MovieListing } from './components/movieListing/MovieListing';
import { RootStore } from './models/RootStore';
import { Provider } from "mobx-react";
jest.mock('./api/MoviesApi');


describe('Component tests ', async () => {

  let store;

  beforeEach( async () => {
    store = RootStore.create();
  });

  it('should render correctly App cmp', async () => {
    const tree = renderer
      .create(
        <Provider rootStore={store}>
          <App />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly GenreSelector cmp', async () => {
    await store.loadMovies();
    await store.loadGenres();
    const tree = renderer
      .create(
        <GenreSelector genres={store.genres} onChange={() => {}}/>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly MovieListing cmp', async () => {
    await store.loadMovies();
    await store.loadGenres();
    const tree = renderer
      .create(
        <MovieListing movies={store.moviesFilteredByrating} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
