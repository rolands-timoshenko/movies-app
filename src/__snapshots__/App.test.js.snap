// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component tests  should render correctly App cmp 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Now playing movies
    </h1>
  </header>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col"
        style={
          Object {
            "marginBottom": 40,
            "padding": 20,
            "textAlign": "left",
          }
        }
      >
        <div
          className="rc-slider rc-slider-with-marks"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "30%",
              }
            }
          />
          <div
            className="rc-slider-step"
          >
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "0%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "10%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "20%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "30%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "40%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "50%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "60%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "70%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "80%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "90%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "100%",
                }
              }
            />
          </div>
          <div
            aria-disabled={false}
            aria-valuemax={10}
            aria-valuemin={0}
            aria-valuenow={3}
            className="rc-slider-handle"
            role="slider"
            style={
              Object {
                "left": "30%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          >
            <span
              className="rc-slider-mark-text rc-slider-mark-text-active"
              style={
                Object {
                  "left": "0%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              0
            </span>
            <span
              className="rc-slider-mark-text rc-slider-mark-text-active"
              style={
                Object {
                  "left": "10%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              1
            </span>
            <span
              className="rc-slider-mark-text rc-slider-mark-text-active"
              style={
                Object {
                  "left": "20%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              2
            </span>
            <span
              className="rc-slider-mark-text rc-slider-mark-text-active"
              style={
                Object {
                  "left": "30%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              3
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "40%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              4
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "50%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              5
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "60%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              6
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "70%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              7
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "80%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              8
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "90%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              9
            </span>
            <span
              className="rc-slider-mark-text"
              style={
                Object {
                  "left": "100%",
                  "marginLeft": "-4.5%",
                  "width": "9%",
                }
              }
            >
              10
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    />
    <div
      className="row"
    />
  </div>
</div>
`;

exports[`Component tests  should render correctly GenreSelector cmp 1`] = `
Array [
  <div
    className="col-6 col-sm-4 col-md-3 col-lg-2"
>
    <button
        className="btn btn-secondary"
        onClick={[Function]}
        style={
            Object {
                "marginBottom": 20,
                "width": "100%",
              }
        }
        type={undefined}
    >
        Action
    </button>
</div>,
  <div
    className="col-6 col-sm-4 col-md-3 col-lg-2"
>
    <button
        className="btn btn-secondary"
        onClick={[Function]}
        style={
            Object {
                "marginBottom": 20,
                "width": "100%",
              }
        }
        type={undefined}
    >
        Adventure
    </button>
</div>,
  <div
    className="col-6 col-sm-4 col-md-3 col-lg-2"
>
    <button
        className="btn btn-secondary"
        onClick={[Function]}
        style={
            Object {
                "marginBottom": 20,
                "width": "100%",
              }
        }
        type={undefined}
    >
        Western
    </button>
</div>,
]
`;

exports[`Component tests  should render correctly MovieListing cmp 1`] = `
Array [
  <div
    className="col-12 col-sm-6 col-md-4 col-lg-3"
>
    <div
        style={
            Object {
                "marginBottom": 20,
              }
        }
    >
        <div
            className="card"
        >
            <img
                alt="Blaze"
                className="card-img-top"
                src="undefined/8Sxf0l5iAFau1sjOzZuk5VX1fcC.jpg"
                width="100%"
            />
            <div
                className="card-body"
            >
                <h5
                    className="card-title"
                >
                    Blaze
                </h5>
                <h6
                    className="card-subtitle"
                >
                    Action, Adventure, Western
                </h6>
                <br />
                <button
                    className="btn btn-secondary"
                    onClick={[Function]}
                    type={undefined}
                >
                    Watch
                </button>
            </div>
        </div>
    </div>
</div>,
  <div
    className="col-12 col-sm-6 col-md-4 col-lg-3"
>
    <div
        style={
            Object {
                "marginBottom": 20,
              }
        }
    >
        <div
            className="card"
        >
            <img
                alt="Book Club"
                className="card-img-top"
                src="undefined/6ORGBbOA45pfKQXvntPf46NoFGO.jpg"
                width="100%"
            />
            <div
                className="card-body"
            >
                <h5
                    className="card-title"
                >
                    Book Club
                </h5>
                <h6
                    className="card-subtitle"
                >
                    Action
                </h6>
                <br />
                <button
                    className="btn btn-secondary"
                    onClick={[Function]}
                    type={undefined}
                >
                    Watch
                </button>
            </div>
        </div>
    </div>
</div>,
  <div
    className="col-12 col-sm-6 col-md-4 col-lg-3"
>
    <div
        style={
            Object {
                "marginBottom": 20,
              }
        }
    >
        <div
            className="card"
        >
            <img
                alt="12 Strong"
                className="card-img-top"
                src="undefined/j18021qCeRi3yUBtqd2UFj1c0RQ.jpg"
                width="100%"
            />
            <div
                className="card-body"
            >
                <h5
                    className="card-title"
                >
                    12 Strong
                </h5>
                <h6
                    className="card-subtitle"
                >
                    Action, Adventure
                </h6>
                <br />
                <button
                    className="btn btn-secondary"
                    onClick={[Function]}
                    type={undefined}
                >
                    Watch
                </button>
            </div>
        </div>
    </div>
</div>,
]
`;
